quinn-runtime/src/lib.rs

```diff
@@ -0,0 +1,71 @@
+#[cfg(feature = "runtime-tokio")]
+mod tokio_runtime;
+#[cfg(feature = "runtime-tokio")]
+pub use tokio_runtime::*;
+
+#[cfg(feature = "runtime-async-std")]
+mod async_std_runtime;
+#[cfg(feature = "runtime-async-std")]
+pub use async_std_runtime::*;
+
+use std::fmt::Debug;
+use std::future::Future;
+use std::io;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use std::time::Instant;
+
+pub trait AsyncWrappedUdpSocket: Sized + Send + Debug {
+    fn new(t: std::net::UdpSocket) -> io::Result<Self>;
+
+    // On Unix we expect to be able to access the underlying std UdpSocket
+    // to be able to implement more advanced features
+    #[cfg(unix)]
+    fn poll_read<T>(
+        &self,
+        f: impl FnOnce(&std::net::UdpSocket) -> io::Result<T>,
+        cx: &mut Context,
+    ) -> Poll<io::Result<T>>;
+    #[cfg(unix)]
+    fn poll_write<T>(
+        &mut self,
+        f: impl FnOnce(&std::net::UdpSocket) -> io::Result<T>,
+        cx: &mut Context,
+    ) -> Poll<io::Result<T>>;
+    #[cfg(unix)]
+    fn get_ref(&self) -> &std::net::UdpSocket;
+
+    // On Non-Unix platforms we only expect to be able to do basic
+    // send_to / recv_from operations.
+    #[cfg(not(unix))]
+    fn poll_recv_from(
+        &self,
+        cx: &mut Context,
+        buf: &mut [u8],
+    ) -> Poll<io::Result<(usize, std::net::SocketAddr)>>;
+    #[cfg(not(unix))]
+    fn poll_send_to(
+        &mut self,
+        cx: &mut Context,
+        buf: &[u8],
+        target: std::net::SocketAddr,
+    ) -> Poll<io::Result<usize>>;
+    #[cfg(not(unix))]
+    fn local_addr(&self) -> io::Result<std::net::SocketAddr>;
+}
+
+pub trait AsyncTimer: Sized + Send + Debug {
+    fn new(i: Instant) -> Self;
+    fn reset(self: Pin<&mut Self>, i: Instant);
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<()>;
+}
+
+pub trait Runtime: Send + 'static {
```

Comment:
I believe the API churn and ergonomic cost to this change would be dramatically reduced if this trait were made object-safe (e.g. replace the associated types with methods on `&self` that return trait objects) and handled in a type-erased manner inside `quinn`.

Reply:
I don't fully understand this proposal, how would the user specify the runtime with this?

Reply:
Would that mean storing a `Box<dyn AsyncWrappedUdpSocket>` instead of a `RT::AsyncWrappedUdpSocket`?

Reply:
If I try to make AsyncWrappedUdpSocket object-safe, I also have to change the `impl FnOnce ... ` to an `Box<dyn FnOnce ...>`.
I can create a third PR for this strategy, but it has a runtime performance cost (~15%) and I don't think we want that.

Reply:
#1346

Reply:
Thanks! You're correct that dynamic dispatch isn't free, but it also isn't very expensive. We must also weigh the cost of impact on compile time (monomorphization undermines separate compilation) and ergonomics (piping type variables around everywhere is a pain). I do not expect that the difference here is measurable, especially compared to the cost of encryption and I/O, but you can run benchmarks to confirm.

Reply:
> I also have to change the impl FnOnce ... to an Box<dyn FnOnce ...>.

I suspect this can be avoided with careful API design. For example, the use of callbacks could be dropped in favor of returning information sufficient to allow the caller to run appropriate code directly.

Reply:
> the use of callbacks could be dropped in favor of returning information sufficient to allow the caller to run appropriate code directly.

This will be hard because _after_ the code is run tokio makes some decisions based on if the recv/send attempt was successful or if it returned WouldBlock. tokio's interface for this is the `guard.try_io()` which also takes a FnOnce.

Reply:
Actually I found a way to do this, will push to the other PR shortly